### PR TITLE
Surface Normal Features: outward normal direction per surface node (+2 channels)

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The model currently has no explicit encoding of the **local surface orientation** at each surface node. It has curvature (scalar) and DSDF (distance to surface), but not the direction the surface is FACING at each point.

Surface normal direction (nx, ny) directly determines the local pressure: in inviscid flow, `p = p_stagnation - 0.5*rho*(V·n)^2` at the surface. The normal tells the model **which direction the surface faces relative to the freestream** — this is the primary geometric quantity that determines pressure.

For OOD generalization (NACA0012 → NACA6416):
- NACA0012 (symmetric): normals are symmetric about the chord line
- NACA6416 (cambered): normals are **asymmetric** — upper surface normals tilt differently from lower surface due to 6% camber and 16% thickness distribution
- The surface normal captures this camber asymmetry in a direct, geometry-invariant way

This is NOT derivable from existing features:
- **DSDF** encodes distance TO the surface, not the surface's own orientation
- **Curvature** is the RATE of change of the normal (d²y/ds²), not the normal itself
- **TE/LE coord frames** encode distance FROM reference points, not local geometry

**Physical motivation:** Every panel method and BEM solver uses surface normals as fundamental inputs. They determine pressure loading, force decomposition, and wake boundary conditions. Adding normals as input features aligns the model's representation with how aerodynamicists parameterize surface pressure.

**Expected improvement:** -1 to -3% p_tan (OOD camber), -1 to -2% p_in (better pressure loading). Zero architecture change, +2 input channels.

## Instructions

All changes in `cfd_tandemfoil/train.py`.

### Step 1: Add config flag

```python
surface_normal_features: bool = False   # outward surface normal (nx, ny) per surface node (+2 channels)
```

### Step 2: Implement `compute_surface_normals`

```python
def compute_surface_normals(raw_xy, is_surface, saf_norm, k=5):
    """Compute outward-pointing surface normal for each surface node.
    
    For each surface node: find k nearest surface neighbors ON THE SAME FOIL,
    fit a local tangent direction from the neighbor displacements,
    rotate 90° to get the normal, then orient outward (away from foil centroid).
    
    Args:
        raw_xy:     [B, N, 2] raw x, y coordinates
        is_surface: [B, N] bool
        saf_norm:   [B, N] saf channel norm (<=0.005 = foil-1, >0.005 = foil-2)
        k: number of nearest neighbors for tangent estimation
    
    Returns: [B, N, 2] (nx, ny) unit normal. Zero for volume nodes.
    """
    B, N = raw_xy.shape[:2]
    result = torch.zeros(B, N, 2, device=raw_xy.device, dtype=raw_xy.dtype)
    
    for b in range(B):
        for foil_idx, threshold_fn in enumerate([
            lambda sn: sn <= 0.005,   # fore-foil
            lambda sn: sn > 0.005,    # aft-foil
        ]):
            mask = is_surface[b] & threshold_fn(saf_norm[b])
            if mask.sum() < 3:
                continue
            
            indices = mask.nonzero(as_tuple=True)[0]  # surface node indices
            xy = raw_xy[b, indices]  # [M, 2]
            centroid = xy.mean(dim=0)  # foil centroid
            
            # For each surface node, find k nearest neighbors
            # Use pairwise distances
            dists = torch.cdist(xy.unsqueeze(0), xy.unsqueeze(0)).squeeze(0)  # [M, M]
            _, knn_idx = dists.topk(min(k+1, len(indices)), dim=1, largest=False)  # [M, k+1]
            knn_idx = knn_idx[:, 1:]  # exclude self
            
            # Local tangent from neighbor displacements
            for i in range(len(indices)):
                neighbors = xy[knn_idx[i]]  # [k, 2]
                displacements = neighbors - xy[i]  # [k, 2]
                # Tangent = principal direction of displacements (SVD or simple average)
                # Simple: weighted average of displacement directions
                d_norms = displacements.norm(dim=1, keepdim=True).clamp(min=1e-8)
                d_unit = displacements / d_norms
                tangent = d_unit.mean(dim=0)  # average direction
                tangent = tangent / tangent.norm().clamp(min=1e-8)
                
                # Normal = rotate tangent 90° CCW
                nx = -tangent[1]
                ny = tangent[0]
                
                # Orient outward: normal should point AWAY from centroid
                to_centroid = centroid - xy[i]
                if (nx * to_centroid[0] + ny * to_centroid[1]) > 0:
                    nx, ny = -nx, -ny  # flip to point outward
                
                result[b, indices[i], 0] = nx
                result[b, indices[i], 1] = ny
    
    return result
```

**IMPORTANT: torch.compile compatibility.** The nested loops above are NOT compile-friendly. You have two options:
1. **Precompute before compile**: compute normals once per batch OUTSIDE the compiled model forward, then pass as additional input features.
2. **Vectorize**: replace the per-node loop with batched operations (cdist + topk are already vectorized; the tangent computation can be vectorized over all nodes simultaneously).

Option 1 is simpler and safer. Compute the normals in the training loop before calling the model, then concatenate them to the input features.

### Step 3: Update input dimension

Add `+ (2 if cfg.surface_normal_features else 0)` to `fun_dim` or `n_x`.

### Step 4: Compute and append in training/eval loops

Before normalization, compute normals from raw coordinates. Then after feature concatenation:
```python
if cfg.surface_normal_features:
    normals = compute_surface_normals(_raw_xy, is_surface, _raw_saf_norm)
    x = torch.cat([x, normals], dim=-1)
```

Apply in **all 4 loops**.

### Step 5: Run 2 seeds

```bash
# Seed 42
cd cfd_tandemfoil && python train.py \
  --agent alphonse --wandb_name "alphonse/surface-normals-s42" \
  --wandb_group "round17/surface-normal-features" \
  --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --surface_normal_features

# Seed 73 — identical but --seed 73 --wandb_name "alphonse/surface-normals-s73"
```

### Step 6: Report results

Table: p_in, p_oodc, p_tan, p_re for both seeds, 2-seed avg, comparison to baseline, W&B run IDs.

## Baseline

Current best (PR #2213, Wake Deficit Feature, 2-seed average):

| Metric | Baseline | Target to beat |
|--------|----------|----------------|
| p_in   | **11.979** | < 11.98 |
| p_oodc | **7.643**  | < 7.65  |
| **p_tan** | **28.341** | **< 28.34** |
| p_re   | **6.300**  | < 6.30  |

W&B runs: hgml7i2r (seed 42), qic03vrg (seed 73)

**Reproduce baseline:**
```bash
cd cfd_tandemfoil && python train.py --agent alphonse --wandb_name "alphonse/baseline-wake-deficit" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT alphonse:

## Results

### Summary

Surface normal features (+2 channels of outward-pointing (nx, ny) per surface node) did **not improve** over baseline. All 4 metrics regressed compared to the baseline (Wake Deficit Feature, PR #2213).

### Metrics Table

| Metric | Baseline | Seed 42 | Seed 73 | 2-seed avg | vs Baseline |
|--------|----------|---------|---------|------------|-------------|
| **p_in**  | 11.979 | 12.833 | 12.369 | **12.601** | +0.622 (worse) |
| **p_oodc** | 7.643 | 8.431 | 7.929 | **8.180** | +0.537 (worse) |
| **p_tan** | 28.341 | 29.247 | 29.062 | **29.155** | +0.814 (worse) |
| **p_re**  | 6.300 | 6.407 | 6.669 | **6.538** | +0.238 (worse) |

Metric notation: p_in = val_in_dist mae_surf_p, p_oodc = val_ood_cond mae_surf_p, p_tan = val_tandem_transfer mae_surf_p, p_re = val_ood_re mae_surf_p. Lower is better.

### Full Surface MAE at Best Epoch (val_in_dist)

| | Seed 42 | Seed 73 |
|--|---------|---------|
| mae_surf_Ux | 2.880 | 2.671 |
| mae_surf_Uy | 0.883 | 0.792 |
| **mae_surf_p** | **12.833** | **12.369** |
| mae_vol_Ux | 0.552 | 0.503 |
| mae_vol_Uy | 0.200 | 0.196 |
| mae_vol_p | 13.102 | 12.571 |

### Run Details

| | Seed 42 | Seed 73 |
|--|---------|---------|
| W&B run | `lkje1j6x` | `0qpde0kn` |
| Best epoch | 143 / 500 (wall-clock timeout) | 144 / 500 (wall-clock timeout) |
| Peak VRAM | ~46.3 GB | ~46.1 GB |

### Command

```bash
# Seed 42
cd cfd_tandemfoil && CUDA_VISIBLE_DEVICES=0 python train.py \
  --agent alphonse --wandb_name "alphonse/surface-normals-s42" \
  --wandb_group "round17/surface-normal-features" --seed 42 \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 160 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature --surface_normal_features
# Seed 73 identical with --seed 73 --wandb_name "alphonse/surface-normals-s73"
```

### What Happened

The surface normals regressed all four metrics by a meaningful margin (+0.6 on p_in, +0.5 on p_oodc, +0.8 on p_tan). This is worse than I expected given the physical motivation.

**Why it likely failed:**

1. **The normals are redundant given existing features.** The model already has DSDF (a per-foil signed distance field with direction information), curvature, and the TE coordinate frame — which together encode local surface orientation quite richly. Adding explicit (nx, ny) may not provide new signal, but does add parameters (and noise) to the model's first-layer projection.

2. **kNN tangent estimation is noisy near high-curvature regions.** Leading edges, trailing edges, and corner points get poor tangent estimates because the k=5 nearest neighbors span a curve that isn't locally linear. This injects noise into the feature.

3. **The vectorized tangent computation (mean of unit displacements) is sensitive to symmetric neighborhoods.** At the LE/TE with a symmetric geometry, the mean cancels out, giving a near-zero tangent and a poorly-defined normal. Orientation flipping via the centroid heuristic may compound errors at these points.

4. **No normalization of the normals beyond unit-length.** The normals are unit vectors (values in [-1, 1]), while other features (DSDF, curvature) go through the learned normalization stats. The new channels have 0 mean and small variance mismatch, which the model can compensate for, but adds early-training instability.

### Bug Fixed

Visualization loop was missing `or cfg.surface_normal_features` in the `_need_te_raw_vis` condition and was not concatenating surface normals before feeding inputs to the visualization model. This caused the visualization to fail with a shape mismatch. Fixed in this PR.

### Suggested Follow-ups

- **Normal-frame velocity projection**: rather than raw (nx, ny), compute the normal-velocity and tangential-velocity components of the freestream (dot product with incoming flow direction) — this is a more directly physics-motivated feature than the raw normal direction.
- **Edge-aware kNN**: exclude "across-the-gap" neighbors when computing tangents for aft-foil nodes (a small chord-length threshold would prevent wrap-around errors at the TE).
- **Simpler curvature from normals**: if we already compute per-node normals, the change in normal direction between adjacent nodes is a better curvature estimate than the existing DSDF gradient curvature — could replace rather than augment.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2229 because ADVISOR: Closing as dead end. All metrics regressed significantly: p_in +5.2%, p_oodc +7.0%, p_tan +2.9%, p_re +3.8%.

Your analysis is excellent. The key failure modes:
1. DSDF gradients already encode surface orientation — (nx, ny) adds no new signal
2. kNN tangent estimation breaks down at high-curvature regions (LE/TE) — exactly where pressure is most sensitive
3. The centroid-based outward orientation heuristic compounds noise

This is the 5th consecutive feature engineering failure this round (after chord-camber, Re-walldist, arc-length PE, domain-split SRF). Strong evidence that the feature engineering space is saturated — DSDF + TE coord frame + wake deficit capture the geometric information the model can use. 

Your suggestion about normal-frame velocity projection (dot(n, V_freestream)) is interesting but suffers from the same noisy normal estimation. Moving you to a different category entirely.

Bug fix on the visualization loop is appreciated — I'll cherry-pick that if needed..
